backends/memcheck/ceed-memcheck-vector.c

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include <ceed/backend.h>
+#include <ceed/ceed.h>
+#include <math.h>
+#include <string.h>
+#include <valgrind/memcheck.h>
+
+#include "ceed-memcheck.h"
+
+//------------------------------------------------------------------------------
+// Has Valid Array
+//------------------------------------------------------------------------------
+static int CeedVectorHasValidArray_Memcheck(CeedVector vec, bool *has_valid_array) {
+  CeedVector_Memcheck *impl;
+  CeedCallBackend(CeedVectorGetData(vec, &impl));
+
+  *has_valid_array = !!impl->array;
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Check if has borrowed array of given type
+//------------------------------------------------------------------------------
+static inline int CeedVectorHasBorrowedArrayOfType_Memcheck(const CeedVector vec, CeedMemType mem_type, bool *has_borrowed_array_of_type) {
+  CeedVector_Memcheck *impl;
+  CeedCallBackend(CeedVectorGetData(vec, &impl));
+  Ceed ceed;
+  CeedCallBackend(CeedVectorGetCeed(vec, &ceed));
+
+  switch (mem_type) {
+    case CEED_MEM_HOST:
+      *has_borrowed_array_of_type = !!impl->array_borrowed;
+      break;
+    default:
+      // LCOV_EXCL_START
+      return CeedError(ceed, CEED_ERROR_BACKEND, "Can only set HOST memory for this backend");
+      // LCOV_EXCL_STOP
+      break;
+  }
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Vector Set Array
+//------------------------------------------------------------------------------
+static int CeedVectorSetArray_Memcheck(CeedVector vec, CeedMemType mem_type, CeedCopyMode copy_mode, CeedScalar *array) {
+  CeedVector_Memcheck *impl;
+  CeedCallBackend(CeedVectorGetData(vec, &impl));
+  CeedSize length;
+  CeedCallBackend(CeedVectorGetLength(vec, &length));
+  Ceed ceed;
+  CeedCallBackend(CeedVectorGetCeed(vec, &ceed));
+
+  if (mem_type != CEED_MEM_HOST) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_BACKEND, "Can only set HOST memory for this backend");
+    // LCOV_EXCL_STOP
+  }
+
+  CeedCallBackend(CeedFree(&impl->array_allocated));
+  CeedCallBackend(CeedFree(&impl->array_owned));
+  switch (copy_mode) {
+    case CEED_COPY_VALUES:
+      CeedCallBackend(CeedCalloc(length, &impl->array_owned));
+      impl->array_borrowed = NULL;
+      impl->array          = impl->array_owned;
+      if (array) memcpy(impl->array, array, length * sizeof(array[0]));
+      else {
+        for (CeedInt i = 0; i < length; i++) impl->array[i] = NAN;
```

Comment:
Instead of blithly Callocing an array full of nice zeros, here we initialize to NAN if the array data is not invalid.